third_party/move/tools/move-asm/src/module_builder.rs

```diff
@@ -106,55 +111,81 @@ impl Display for QualifiedId {
+    /// builder is for a script, `module_id_opt` should be `None`, otherwise
+    /// contain the module id.
     pub fn new(
         options: ModuleBuilderOptions,
         context_modules: impl IntoIterator<Item = &'a CompiledModule>,
         module_id_opt: Option<&ModuleId>,
     ) -> Self {
-        let mut module = CompiledModule {
-            version: options.bytecode_version,
-            self_module_handle_idx: ModuleHandleIndex(0),
-            ..Default::default()
-        };
-
-        module.module_handles.push(ModuleHandle {
-            address: AddressIdentifierIndex(0),
-            name: IdentifierIndex(0),
-        });
-        let ModuleId {
-            address: script_address,
-            name: script_name,
-        } = language_storage::pseudo_script_module_id().clone();
-        let (address, name) = if let Some(mid) = module_id_opt {
-            assert_ne!(mid.address, script_address, "address reserved for scripts");
-            (mid.address, mid.name.clone())
-        } else {
-            (script_address, script_name)
-        };
-        module.address_identifiers.push(address);
-        module.identifiers.push(name);
-
         let context_modules = context_modules
             .into_iter()
             .map(|m| (ModuleView::new(m).id(), m))
             .collect();
-        Self {
-            module: RefCell::new(module),
-            options,
-            context_modules,
-            ..Default::default()
+        if let Some(mid) = module_id_opt {
+            let mut module = CompiledModule {
+                version: options.bytecode_version,
+                self_module_handle_idx: ModuleHandleIndex(0),
+                ..Default::default()
+            };
+            module.module_handles.push(ModuleHandle {
+                address: AddressIdentifierIndex(0),
+                name: IdentifierIndex(0),
+            });
+            module.address_identifiers.push(mid.address);
+            module.identifiers.push(mid.name.clone());
+            let builder = Self {
+                module: RefCell::new(module),
+                options,
+                context_modules,
+                ..Default::default()
+            };
+            builder
+                .module_to_idx
+                .borrow_mut()
+                .insert(mid.clone(), ModuleHandleIndex::new(0));
+            builder
+                .address_to_idx
+                .borrow_mut()
+                .insert(*mid.address(), AddressIdentifierIndex::new(0));
+            builder
+                .name_to_idx
+                .borrow_mut()
+                .insert(mid.name().to_owned(), IdentifierIndex::new(0));
+            builder
+        } else {
+            // Use a pseudo handle index for scripts
+            let module = CompiledModule {
+                version: options.bytecode_version,
+                self_module_handle_idx: ModuleHandleIndex(TableIndex::MAX),
```

Comment:
Maybe you can have a function that defines this as `fn script_handle_idx() -> ModuleHandleIndex{ ModuleHandleIndex(TableIndex::MAX) }` so that we do not use max index in multiple places in case we change something later.